### PR TITLE
MAINT: Start applying pycodestyle error rules (E)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
In preparation for https://github.com/numpy/numpy/issues/24994.

### CODE REVIEW

**Comment by charris** [comment]
Needs rebase.

**Comment by charris** [comment]
Might want tag ignore some of the long lines.

**Comment by charris** [comment]
@DimitriPapadopoulos Are you intentionally keeping this a draft?

**Comment by DimitriPapadopoulos** [comment]
I need to apply more `E` pycodestyle rules and thought you would prefer a single PR. Actually I prefer smaller changes, so I will remove the Draft label soon and leave other changes for later.

File: doc/source/conf.py [hunk 1]
```diff
@@ -306,19 +305,19 @@ def setup(app):
 # (source start file, target name, title, author, document class [howto/manual]).
 _stdauthor = 'Written by the NumPy community'
 latex_documents = [
-  ('reference/index', 'numpy-ref.tex', 'NumPy Reference',
-   _stdauthor, 'manual'),
-  ('user/index', 'numpy-user.tex', 'NumPy User Guide',
-   _stdauthor, 'manual'),
+    ('reference/index', 'numpy-ref.tex', 'NumPy Reference',
+    _stdauthor, 'manual'),
```
Comment (charris) [inline doc/source/conf.py:309]: Shouldn't the continuation have an extra space?

File: numpy/_core/_dtype.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ def _scalar_str(dtype, short):
         else:
             return "'%sU%d'" % (byteorder, dtype.itemsize / 4)
 
-    elif dtype.type == str:
+    elif dtype.type is str:
```
Comment (charris) [inline numpy/_core/_dtype.py:129]: Hmm, this might change behavior. @seberg Thoughts?

File: numpy/_core/code_generators/generate_umath.py [hunk 3]
```diff
@@ -259,9 +259,9 @@ def english_upper(s):
...
 
 
-#each entry in defdict is a Ufunc object.
+# each entry in defdict is a Ufunc object.
 
-#name: [string of chars for which it is defined,
+# name: [string of chars for which it is defined,
 #       string of characters using func interface,
 #       tuple of strings giving funcs for data,
 #       (in, out), or (instr, outstr) giving the signature as character codes,
```
Comment (charris) [inline numpy/_core/code_generators/generate_umath.py:267]: Might add an extra space here even if a comment.

File: numpy/_core/tests/test_casting_unittests.py [hunk 4]
```diff
@@ -267,8 +268,8 @@ def test_simple_cancast(self, from_Dt):
                 for to_dt in [to_Dt(), to_Dt().newbyteorder()]:
                     casting, (from_res, to_res), view_off = (
                             cast._resolve_descriptors((from_dt, to_dt)))
-                    assert(type(from_res) == from_Dt)
-                    assert(type(to_res) == to_Dt)
+                    assert type(from_res) is from_Dt
```
Comment (charris) [inline numpy/_core/tests/test_casting_unittests.py:271]: Might want to check that this is correct.

File: numpy/_core/tests/test_datetime.py [hunk 5]
```diff
@@ -1154,26 +1154,26 @@ def test_datetime_add(self):
...
+                        # NumPy scalars
+                        (np.datetime64('2012-12-21', '[D]'),
+                         np.datetime64('2012-12-24', '[D]'),
+                         np.datetime64('1940-12-24', '[D]'),
+                         np.datetime64('1940-12-24T00', '[h]'),
+                         np.datetime64('1940-12-23T13', '[h]'),
+                         np.datetime64('NaT', '[D]'),
+                         np.timedelta64(3, '[D]'),
+                         np.timedelta64(11, '[h]'),
+                         np.timedelta64(3*24 - 11, '[h]'))]:
```
Comment (charris) [inline numpy/_core/tests/test_datetime.py:1176]: Hmm, I think here and below the contents are over indented. Might be a better way to write these, maybe put the opening `[` on the preceding line and dispense with the line continuation?

File: numpy/_core/tests/test_numeric.py [hunk 6]
```diff
@@ -3501,7 +3504,7 @@ def test_dtype_str_bytes(self, likefunc, dtype):
         b = a[:, ::2]  # Ensure b is not contiguous.
         kwargs = {'fill_value': ''} if likefunc == np.full_like else {}
         result = likefunc(b, dtype=dtype, **kwargs)
-        if dtype == str:
+        if dtype is str:
```
Comment (charris) [inline numpy/_core/tests/test_numeric.py:3507]: Might be worth a check.

File: numpy/_core/tests/test_shape_base.py [hunk 7]
```diff
@@ -686,9 +685,9 @@ def test_nested(self, block):
         result = block([
             [
                 block([
-                   [one],
-                   [three],
-                   [four]
+                    [one],
+                    [three],
+                    [four]
                 ]),
```
Comment (charris) [inline numpy/_core/tests/test_shape_base.py:691]: This nesting here is difficult to understand, could use a reformatting.

File: numpy/_core/tests/test_simd.py [hunk 8]
```diff
@@ -185,7 +185,8 @@ def test_operators_logical(self):
         assert data_xnor == vxnor
 
     def test_tobits(self):
-        data2bits = lambda data: sum([int(x != 0) << i for i, x in enumerate(data, 0)])
+        def data2bits(data):
+            return sum([int(x != 0) << i for i, x in enumerate(data, 0)])
```
Comment (charris) [inline numpy/_core/tests/test_simd.py:189]: I'm a bit surprised a blank line wasn't added here.

File: numpy/_core/tests/test_simd.py [hunk 9]
```diff
@@ -416,7 +418,8 @@ def test_sqrt(self):
             sqrt  = self.sqrt(self.setall(case))
             assert sqrt == pytest.approx(data_sqrt, nan_ok=True)
 
-        data_sqrt = self.load([math.sqrt(x) for x in data]) # load to truncate precision
+        data_sqrt = self.load([math.sqrt(x)  # load to truncate precision
+                               for x in data])
```
Comment (charris) [inline numpy/_core/tests/test_simd.py:422]: Would be nicer to move the training comment to the preceding line.

File: numpy/_typing/_array_like.py [hunk 10]
```diff
@@ -39,7 +39,8 @@
 # any and all remaining overloads
 @runtime_checkable
 class _SupportsArray(Protocol[_DType_co]):
-    def __array__(self) -> ndarray[Any, _DType_co]: ...
+    def __array__(self) -> ndarray[Any, _DType_co]:
+        ...
```
Comment (charris) [inline numpy/_typing/_array_like.py:43]: I think all these typing fixes should be avoided. @jorenham Thoughts?

File: numpy/f2py/_src_pyf.py [hunk 11]
```diff
@@ -143,7 +145,7 @@ def listrepl(mobj):
             lnames[name] = thelist
         return "<%s>" % name
 
-    substr = list_re.sub(listrepl, substr) # convert all lists to named templates
+    substr = list_re.sub(listrepl, substr)  # convert all lists to named templates
                                            # newnames are constructed as needed
```
Comment (charris) [inline numpy/f2py/_src_pyf.py:149]: Might want to line this up. Maybe better, put the comments on the preceding line. I hate trailing comments :)

File: numpy/f2py/_src_pyf.py [hunk 12]
```diff
@@ -197,11 +199,12 @@ def process_str(allstr):
...
         names.update(defs)
         writestr += expand_sub(newstr[sub[0]:sub[1]], names)
-        oldend =  sub[1]
+        oldend = sub[1]
     writestr += newstr[oldend:]
 
     return writestr
 
+
 include_src_re = re.compile(r"(\n|\A)\s*include\s*['\"](?P<name>[\w\d./\\]+\.src)['\"]", re.I)
```
Comment (charris) [inline numpy/f2py/_src_pyf.py:208]: This line seems long ...

File: numpy/linalg/tests/test_linalg.py [hunk 13]
```diff
@@ -1066,7 +1066,7 @@ def tz(mat):
 
         for mat in self.rshft_all:
             tz(mat.astype(dt))
-            if dt != object:
+            if dt != object:  # noqa: E721
```
Comment (charris) [inline numpy/linalg/tests/test_linalg.py:1069]: I'm curious why `dt is not object` didn't work here.

File: numpy/ma/core.py [hunk 14]
```diff
@@ -8264,8 +8283,8 @@ def _convolve_or_correlate(f, a, v, mode, propagate_mask):
     if propagate_mask:
         # results which are contributed to by either item in any pair being invalid
         mask = (
-            f(getmaskarray(a), np.ones(np.shape(v), dtype=bool), mode=mode)
-          | f(np.ones(np.shape(a), dtype=bool), getmaskarray(v), mode=mode)
+            f(getmaskarray(a), np.ones(np.shape(v), dtype=bool), mode=mode) |
```
Comment (charris) [inline numpy/ma/core.py:8286]: IIRC, Python changed the style to prefer putting the operator in front. Don't recall if that was for Python or C code.

File: numpy/typing/tests/data/pass/ndarray_misc.py [hunk 15]
```diff
@@ -14,7 +14,9 @@
 import numpy as np
 import numpy.typing as npt
 
-class SubClass(npt.NDArray[np.float64]): ...
+class SubClass(npt.NDArray[np.float64]):
+    ...
+
```
Comment (charris) [inline numpy/typing/tests/data/pass/ndarray_misc.py:19]: Again the typing.

**Comment by charris** [comment]
That was no fun, it would be better to keep the PRs down, and maybe do fewer style changes together. Going through a bunch of lines that add spaces in comments or blank lines is pretty easy, but mixing them in with more complicated fixes slows things down.

**Comment by jorenham** [comment]
For the static typing-related changes I agree with @charris that current "inlined ellipsis style", i.e. `: ...`, is conventional. 
I believe that `ruff format` always enforces this for (at least) `typing.Protocol`, and always does so in `.pyi` stubs.

File: numpy/_typing/_array_like.py (hunk 10 above)
Comment (jorenham) [reply numpy/_typing/_array_like.py:43]: See https://github.com/numpy/numpy/pull/27308#issuecomment-2332596335

Comment (DimitriPapadopoulos) [reply numpy/_typing/_array_like.py:43]: What can I say? I'm just enforcing pycodestyle rules, not my choice. Unlike ruff, pycodestyle appears to be very opinionated, in the sense that errors cannot be silenced with `noqa`.

File: numpy/ma/core.py (hunk 14 above)
Comment (DimitriPapadopoulos) [reply numpy/ma/core.py:8286]: Indeed, see [W503](https://www.flake8rules.com/rules/W503.html) / [W504](https://www.flake8rules.com/rules/W504.html).

To be honest, I don't care about formatting, just about silencing pycodestyle. I'd rather let a formatter do the formatting for me. However, I understand black or ruff don't work well on numerical code.

File: numpy/_core/tests/test_simd.py (hunk 8 above)
Comment (DimitriPapadopoulos) [reply numpy/_core/tests/test_simd.py:189]: These are automated fixes by ruff, and pycodestyle doesn't have anything to say about it at this point. I was thinking of handling this error in an upcoming PR that would handle the relevant `E` pycodestyle rule.

File: numpy/_core/_dtype.py (hunk 2 above)
Comment (seberg) [reply numpy/_core/_dtype.py:129]: If this was `dtype ==` then yes, as is.  This is fine, although it is a bit funny to not use it above as well.

File: numpy/_typing/_array_like.py (hunk 10 above)
Comment (jorenham) [reply numpy/_typing/_array_like.py:43]: `# fmt: skip` maybe?

Comment (jorenham) [reply numpy/_typing/_array_like.py:43]: ruff also includes pycodestyle btw: https://docs.astral.sh/ruff/rules/#pycodestyle-e-w

File: numpy/ma/core.py (hunk 14 above)
Comment (charris) [reply numpy/ma/core.py:8286]: I'm happy with leaving the change here, just reminded of the preference change.

**Comment by charris** [comment]
@DimitriPapadopoulos What are your plans here? With a few cleanups we could put this in.

**Comment by DimitriPapadopoulos** [comment]
@charris I am afraid I unintentionally deleted this branch 😳 That's not necessarily a big deal because I am not certain any more that fixing Pycodestyle issues in the whole codebase is the best strategy to move to the ruff linter. Perhaps we can continue the discussion about the best strategy in #24994.

**Comment by DimitriPapadopoulos** [comment]
Superseded by #27300.

**Pull request closed without merging**